src/app/Library/CrudPanel/Traits/Validation.php

```diff
@@ -202,42 +202,35 @@ public function validateRequest()
                 return app($formRequest);
             }
 
-            // create an alias of the provided FormRequest so we can create a new class that extends it.
-            // we can't use $variables to extend classes.
-            class_alias(get_class(new $formRequest), 'DeveloperProvidedFormRequest');
-
-            // create a new anonymous class that will extend the provided developer FormRequest
-            // in this class we will merge the FormRequest rules() and messages() with the ones provided by developer in fields.
-            $extendedRequest = new class($rules, $messages) extends \DeveloperProvidedFormRequest
-            {
-                private $_rules;
-                private $_messages;
-
-                public function __construct($rules, $messages)
-                {
-                    parent::__construct();
-                    $this->_rules = $rules;
-                    $this->_messages = $messages;
-                }
-
-                public function rules()
-                {
-                    return array_merge(parent::rules(), $this->_rules);
-                }
-
-                public function messages()
-                {
-                    return array_merge(parent::messages(), $this->_messages);
-                }
-            };
+            $formRequest = (new $formRequest)->createFrom($this->getRequest());
+            $extendedRules = $this->mergeRules($formRequest, $rules);
+            $extendedMessages = array_merge($messages, $formRequest->messages());
 
             // validate the complete request with FormRequest + controller validation + field validation (our anonymous class)
-            return app(get_class($extendedRequest), ['rules' => $rules, 'messages' => $messages]);
+            return $formRequest->validate($extendedRules, $extendedMessages);
         }
 
         return ! empty($rules) ? $this->checkRequestValidity($rules, $messages) : $this->getRequest();
     }
 
+    /**
+     * Return an array containing the request rules and the field/controller rules merged.
+     * The rules in request will take precedence over the ones in controller/fields.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  array  $rules
+     * @return array
+     */
+    private function mergeRules($request, $rules)
+    {
+        $extendedRules = [];
+        foreach ($request->rules() as $ruleKey => $rule) {
+            $extendedRules[$ruleKey] = array_key_exists($ruleKey, $rules) ? $rule.'|'.$rules[$ruleKey] : $rule;
+        }
+
+        return array_merge($rules, $extendedRules);
+    }
```

Comment:
@pxpm you are joining rules from controller and request here 👌

I'm afraid we can't do this 😬
Developers may set their rules in array format.

```php
'text' => ['required', 'min:5', 'max:255'],
```

This generates a `Array to string conversion` 500 error.

We could go and merge them by a pipe `|`, but then there are complex rules, and function rules, we probably don't want to go there 😅

Probably the best is to keep it simple, request will take precedence over the controller and that's it 🤷‍♂️

Reply:
@promatik I agree, I let array definition slip me, thanks. 👍  
But I think we can't give request precedence over controller, they need to be both validated, otherwise the rules you define in your FormRequest like the redirect, prepareForValidation etc would not take effect when validating the controller rules. 

About the problem you reported, maybe the solution is doing it the other way around and ALWAYS use array for validation? So if developer provide a string we explode pipes `|` and add then into validation array ?